File: app/key_algo/bilibili_interface.rb
=begin
  http://www.bilibili.com/video/av4371609/
  
  http://interface.bilibili.com/playurl?cid=7076936&player=1&ts=1460728450&sign=705c096777c7ed41d5cee161913c35be
  
  视频地址转 interface 地址
=end

require 'digest'


cid = '7076936'
# now we get interface url
appkey='85eb6835b0a1034e';  
secretkey = '2ad42749773c441109bdc0191257a664'
sign_this = Digest::MD5.hexdigest('appkey=' + appkey + '&cid=' + cid + secretkey)
interface_url = 'http://interface.bilibili.com/playurl?appkey=' + appkey + '&cid=' + cid + '&sign=' + sign_this  
puts(interface_url)












# md5 = Digest::MD5.new
#md5.update 'message1'
#md5 << '123'                     # << is an alias for update

#puts(md5.hexdigest)                         #=> "94af09c09bb9..."



# 【10分钟速成课：哲学】第5集 - 尼奥，遇见勒内·笛卡尔怀疑主义
# http://www.bilibili.com/video/av4371609/
# cid 7076936
# http://interface.bilibili.com/playurl?cid=7076936&player=1&ts=1460728450&sign=705c096777c7ed41d5cee161913c35be
# 21:54
# 1460728450

# http://interface.bilibili.com/playurl?cid=7076936&player=1&ts=1460728588&sign=9f38b386a9d6eaf0135b5ccab2c474da

# ts = timestamp = 当前请求时间的时间戳
# sign的长度和 md5 一样那应该就是 md5算法了

# 那现在就是折腾看看怎么弄成正确 sign 算法


# cid=7076936&player=1&ts=1460728588&sign=9f38b386a9d6eaf0135b5ccab2c474da

# cid = 7076936
# player = 1
# ts = 1460728588
# sign=9f38b386a9d6eaf0135b5ccab2c474da

#md5 << 'appsecret=2ad42749773c441109bdc0191257a664&'
#md5 << 'cid=7076936'  
#md5 << 'cid=7076936&'                
#md5 << 'player=1'        
#md5 << 'player=1&'              
#md5 << 'ts=1460728588'         
#md5 << 'ts=1460728588'               



#md5 << 'appsecret=ea85624dfcf12d7cc7b2b3a94fac1f2c&appkey=f3bb208b3d081dc8&cid=3885454&'
#sign=3aa2879fb591b4e297ed3e69156c821e 。


#puts(md5.hexdigest)                     
# 9f38b386a9d6eaf0135b5ccab2c474da


#1 搞对签名，然后再搞模拟请求头的事情

File: app/key_algo/bilibili_interface.py
#python3

import sys
import hashlib
#print(hashlib.md5('123'.encode()).hexdigest())

appkey='85eb6835b0a1034e';  
secretkey = '2ad42749773c441109bdc0191257a664'

#cid = '2051663'
#sign_this = hashlib.md5(bytes('appkey=' + appkey + '&cid=' + cid + secretkey, 'utf-8')).hexdigest()  
#url = 'http://interface.bilibili.com/playurl?appkey=' + appkey + '&cid=' + cid + '&sign=' + sign_this  
#print(url)
#============================================================================
# http://interface.bilibili.com/playurl?cid=7076936&player=1&ts=1460728588&sign=9f38b386a9d6eaf0135b5ccab2c474da
# cid = 7076936
# player = 1
# ts = 1460728588
# sign=9f38b386a9d6eaf0135b5ccab2c474da

cid = '7076936'
sign_this = hashlib.md5(bytes('appkey=' + appkey + '&cid=' + cid + secretkey, 'utf-8')).hexdigest()  
#url = 'http://interface.bilibili.com/playurl?appkey=' + appkey + '&cid=' + cid + '&sign=' + sign_this  
#url = 'http://interface.bilibili.com/playurl?cid=' +cid +'&player=1&ts=1460728588' + '&sign=' + sign_this  

print(sign_this)

# output sign should be: 9f38b386a9d6eaf0135b5ccab2c474da
sys.exit()


















APPKEY = '85eb6835b0a1034e'
SECRETKEY = '2ad42749773c441109bdc0191257a664'


#----------------------------------------------------------------------
def calc_sign(string):
    """str/any->str
    return MD5."""
    return str(hashlib.md5(str(string).encode('utf-8')).hexdigest())

#----------------------------------------------------------------------
def get_flash_live_address_from_cid2222(cid):
    """"""
    str_to_hash = 'cid={cid}&player=1&quality=0{SECRETKEY}'.format(cid = cid, SECRETKEY = SECRETKEY)
    url = 'http://live.bilibili.com/api/playurl?cid={cid}&player=1&quality=0&sign={sign}'.format(cid = cid, sign = calc_sign(str_to_hash))
    HEADER = {
        'User-Agent':'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.89 Safari/537.36',
    'Referer': 'http://live.bilibili.com/' + str(cid),
    'X-Requested-With': 'ShockwaveFlash/18.0.0.209'}
    request = urllib2.Request(url)
    response = urllib2.urlopen(request)
    data = response.read()
    #print(data)
    return str(xmltodict.parse(data)['video']['durl']['url'])
    
    


def get_flash_live_address_from_cid(cid):
    str_to_hash = 'cid={cid}&player=1&quality=0{SECRETKEY}'.format(cid = cid, SECRETKEY = SECRETKEY)
    a = 'http://live.bilibili.com/api/playurl?cid={cid}&player=1&quality=0&sign={sign}'.format(cid = cid, sign = calc_sign(str_to_hash))
    return a

print( get_flash_live_address_from_cid("7076936") )
    
